### PR TITLE
Integrating keycard signing in case of buying stickers/ens username

### PR DESCRIPTION
**Pull request by saledjenic**
We should let the client decide where to sign a transaction and once the tx is signed let him submit the tx providing the signature. This PR allows exactly that.

### CODE REVIEW

**Comment by saledjenic** [comment]
@ajayesivan @ilmotta could you please check if the mobile app is using any of the following calls and if yes, could you or somebody else align with the following changes done in this pr:

**Ens Service**

1. Instead of the old `ens_release` call we should do the following calls:
- `ens_prepareTxForReleasingRegisteredEnsUsername`
- decide where to sign the tx from the previous call on a keycard or by calling `ens_signPreparedTx` for regular Status user
- call `ens_sendPreparedTxWithSignature`

2. Instead of the old `ens_register` call we should do the following calls:
- `ens_prepareTxForRegisteringEnsUsername`
- decide where to sign the tx from the previous call on a keycard or by calling `ens_signPreparedTx` for regular Status user
- call `ens_sendPreparedTxWithSignature`

3. Instead of the old `ens_setPubKey` call we should do the following calls:
- `ens_prepareTxForSettingPublicKey`
- decide where to sign the tx from the previous call on a keycard or by calling `ens_signPreparedTx` for regular Status user
- call `ens_sendPreparedTxWithSignature`

**Stickers Service**

1. Instead of the old `stickers_buy` call we should do the following calls:
- `stickers_prepareTxForBuyingStickers`
- decide where to sign the tx from the previous call on a keycard or by calling `stickers_signPreparedTx` for regular Status user
- call `stickers_sendPreparedTxWithSignature`

**Review by alaibe — APPROVED** [review]

**Comment by ilmotta** [comment]
Thank you for the detailed changelog @saledjenic. The mobile repo is definitely affected by the endpoint changes.

I'm going to create an issue in `status-mobile` and see if we have capacity to make the adaptations this week. I'm going to report back here with any findings.

File: services/ens/api.go [hunk 1]
```diff
@@ -336,42 +348,134 @@ func (api *API) Price(ctx context.Context, chainID uint64) (string, error) {
 	return fmt.Sprintf("%x", price), nil
 }
 
-func (api *API) Release(ctx context.Context, chainID uint64, txArgs transactions.SendTxArgs, password string, username string) (string, error) {
-	registryAddr, err := api.usernameRegistrarAddr(ctx, chainID)
+// Signs a transaction using appropriate local keystore file that is decrypted with provided password and returns the signature in hex format.
+// The tx that will be signed is the one that was prepared by the last call to
+// PrepareTxForRegisteringEnsUsername or PrepareTxForSettingPublicKey or PrepareTxForReleasingRegisteredEnsUsername.
+func (api *API) SingPreparedTx(password string) (interface{}, error) {
```
Comment (endulab) [inline services/ens/api.go]: Sing -> Sign

Comment (endulab) [inline services/ens/api.go]: Why do you return interface{} and not string?

File: services/stickers/transactions.go [hunk 2]
```diff
@@ -65,23 +71,99 @@ func (api *API) Buy(ctx context.Context, chainID uint64, txArgs transactions.Sen
...
+		packID:        packID,
+	}
+
+	signer := ethTypes.NewLondonSigner(new(big.Int).SetUint64(api.txSignDetails.chainID))
+	return signer.Hash(api.txSignDetails.txBeingSigned), nil
+}
+
+// Signs a transaction using appropriate local keystore file that is decrypted with provided password and returns the signature in hex format.
+// The tx that will be signed is the one that was prepared by the last call to PrepareTxForBuyingStickers method.
+func (api *API) SingPreparedTx(password string) (interface{}, error) {
```
Comment (endulab) [inline services/stickers/transactions.go]: The code is the same as in ens/api.go. Can you move it to eg. services/utils/utils.go? There is a GetSigner function which you can use also.

File: services/stickers/transactions.go [hunk 3]
```diff
@@ -65,23 +71,99 @@ func (api *API) Buy(ctx context.Context, chainID uint64, txArgs transactions.Sen
...
+	hash := signer.Hash(api.txSignDetails.txBeingSigned)
+
+	signature, err := crypto.Sign(hash[:], selectedAccount.PrivateKey)
+
+	return types.EncodeHex(signature), err
+}
+
+// Sends a transaction using provided signature and returns the transaction hash.
+// Provided signature must be in hex format and must be the result of signing the tx that was prepared by the last call to PrepareTxForBuyingStickers method.
+func (api *API) SendPreparedTxWithSignature(ctx context.Context, signature string) (string, error) {
```
Comment (endulab) [inline services/stickers/transactions.go:103]: Isn't it the same (mostly) as in ens/api.go?

File: services/stickers/transactions.go (hunk 2 above)
Comment (cammellos) [inline services/stickers/transactions.go]: Sign I take?

File: services/ens/api.go (hunk 1 above)
Comment (saledjenic) [reply services/ens/api.go]: While I was developing I used `[]byte`, then `string` and then it remained like that. Np, will change that.

File: services/stickers/transactions.go (hunk 2 above)
Comment (saledjenic) [reply services/stickers/transactions.go]: > There is a GetSigner function which you can use also.

Well, it looks similar, but it doesn't do exactly the same thing as `GetSigner` does. 
- `GetSigner`:
  - makes a tx hash for signing
  - signs tx hash building a signature
  - sends a tx using prepared signature and returns tx of sent tx
- `SingPreparedTx`:
  - signs prepared tx and returns signature
  
> Can you move it to eg. services/utils/utils.go?

Yes, those 2 are very similar, basically they do the same thing, but over different tx. One signs a tx prepared in the context of ens context, other signs a tx prepared in the stickers context. But yes, I can move those few lines, the common part, to the utils.go file, I agree.

**Comment by saledjenic** [comment]
@cammellos @endulab comments addressed, thanks. :)

File: services/stickers/transactions.go (hunk 3 above)
Comment (saledjenic) [reply services/stickers/transactions.go:103]: It is not, different tx, different contract, different context.

**Review by stefandunca — APPROVED** [review]
👍🏼

File: services/stickers/transactions.go [hunk 4]
```diff
@@ -65,23 +71,89 @@ func (api *API) Buy(ctx context.Context, chainID uint64, txArgs transactions.Sen
...
+
+	return utils.SignTx(api.txSignDetails.chainID, api.accountsManager, api.keyStoreDir,
+		types.Address(api.txSignDetails.from), password, api.txSignDetails.txBeingSigned)
+}
+
+// Sends a transaction using provided signature and returns the transaction hash.
+// Provided signature must be in hex format and must be the result of signing the tx that was prepared by the last call to PrepareTxForBuyingStickers method.
+func (api *API) SendPreparedTxWithSignature(ctx context.Context, signature string) (string, error) {
+	if api.txSignDetails == nil {
+		return "", errors.New("no known tx to send to register ens username")
```
Comment (endulab) [inline services/stickers/transactions.go:105]: it is stickers, not ens

File: services/stickers/transactions.go (hunk 3 above)
Comment (endulab) [reply services/stickers/transactions.go:103]: If you look at SendPreparedTxWithSignature in ens service, most of the code is identical, only the end differs.

Comment (endulab) [reply services/stickers/transactions.go:103]: Aaa, sorry, different api.

**Review by endulab — APPROVED** [review]

**Comment by flexsurfer** [comment]
hey @saledjenic on mobile we use `ens_registerPrepareTx` and `ens_setPubKeyPrepareTx` just wondering if we need to change this

**Comment by qfrank** [comment]
Hi @saledjenic , can't we just keep the original implementation for endpoints: `ens_register` and `ens_setPubKey` ? cc @cammellos @flexsurfer

**Comment by flexsurfer** [comment]
@qfrank  ens_register and ens_setPubKey are not used on mobile, why do you need them?

**Comment by qfrank** [comment]
> @qfrank ens_register and ens_setPubKey are not used on mobile, why do you need them?

I thought you see the usages on `ens_register` and `ens_setPubKey` since i saw you removed them in your PR. But they shouldn't be removed otherwise ens registration flow won't function proberbly @flexsurfer

**Comment by flexsurfer** [comment]
> I thought you see the usages on `ens_register` and `ens_setPubKey` since i saw you removed them in your PR

i removed them just because they are not used, we use ens_registerPrepareTx and ens_setPubKeyPrepareTx

**Comment by qfrank** [comment]
> > I thought you see the usages on `ens_register` and `ens_setPubKey` since i saw you removed them in your PR
> 
> i removed them just because they are not used, we use ens_registerPrepareTx and ens_setPubKeyPrepareTx

actually they're used, we should keep them except you don't want mobile ens works as before

**Comment by flexsurfer** [comment]
> actually they're used, we should keep them except you don't want mobile ens works as before

could you please give an example of where they are used, in my mobile PR, because the point of this PR is to replace original methods, they should be used on both desktop and mobile

**Comment by qfrank** [comment]
> > actually they're used, we should keep them except you don't want mobile ens works as before
> 
> could you please give an example of where they are used, in my mobile PR, because the point of this PR is to replace original methods, they should be used on both desktop and mobile

added more review comments in your PR

**Comment by qfrank** [comment]
> hey @saledjenic on mobile we use `ens_registerPrepareTx` and `ens_setPubKeyPrepareTx` just wondering if we need to change this

I think we can remove `ens_registerPrepareTx` and `ens_setPubKeyPrepareTx` @saledjenic ? because we have new implementation `ens_prepareTxForRegisteringEnsUsername` and `ens_prepareTxForSettingPublicKey` now cc @flexsurfer

**Comment by saledjenic** [comment]
@qfrank @flexsurfer sorry for the late reply.

> hey @saledjenic on mobile we use ens_registerPrepareTx and ens_setPubKeyPrepareTx just wondering if we need to change this

@flexsurfer not sure how the mobile app uses them, but since `ens_register` and `ens_setPubKey` do not exist anymore, I am not sure if we need them at all anymore. Maybe we should remove those to as well?

> Hi @saledjenic , can't we just keep the original implementation for endpoints: ens_register and ens_setPubKey ? cc @cammellos @flexsurfer

@qfrank if we need them, we can, but why do we need to have many API endpoints for the same feature? I am more for aligning the desktop and mobile app on this.

> I think we can remove ens_registerPrepareTx and ens_setPubKeyPrepareTx @saledjenic ? because we have new implementation ens_prepareTxForRegisteringEnsUsername and ens_prepareTxForSettingPublicKey now cc @flexsurfer

Exactly, if there are no other places in the mobile app that are using those endpoints I would remove them? 
Could you confirm and let me know if is that ok by your side? @flexsurfer

**Comment by qfrank** [comment]
> if we need them, we can, but why do we need to have many API endpoints for the same feature? I am more for aligning the desktop and mobile app on this.

It depends on how much time we can wait for aligning, the quickest way to keep mobile ens feature work as before is keep the original implementation so that we can align it later after you merge this PR. If time is not what we need concidering, then let's just aligning on this. @saledjenic cc @cammellos @flexsurfer

**Comment by saledjenic** [comment]
Ok, I can revert those endpoints and just add those that are needed for the desktop app and you can align with them later.
Does that work for you better? @flexsurfer

**Comment by flexsurfer** [comment]
hey @saledjenic i think it was already implemented and we have `ens_registerPrepareTx` and `ens_setPubKeyPrepareTx` and i don't understand the point of having `ens_signPreparedTx` and `ens_sendPreparedTxWithSignature`, and same for stickers, why don't we have just general `wallet_signPreparedTx` and `wallet_sendPreparedTxWithSignature`, so on mobile currently it works next way, we prepare transaction and then sign and send it with `sendTransaction` why do we want to change this?

so we have a general sign transaction dialog, which signs and sends any transaction with user password, and adding conditions to it, or having some special dialog for stickers and ens doesn't make sense

**Comment by flexsurfer** [comment]
so on mobile for ens it works next way
1. `ens_registerPrepareTx` we get transaction object
2. sign and send it as any other transaction with a general wallet sign dialog
3. on success `ens_register` for adding a record in the app db and then
4. watch transaction progress

and it works fine, why would we want to change this? it also works great with the keycard, probably desktop team could just use this instead of creating new API?

**Comment by saledjenic** [comment]
> and it works fine, why would we want to change this? it also works great with the keycard, probably desktop team could just use this instead of creating new API?

@flexsurfer if you want I can update the code, leaving functions as they were and just adding those we need for the desktop app. Is that ok?

We need the tx hash that needs to be signed, and I don't see any that returns the tx hash, among the methods you mentioned.

**Comment by flexsurfer** [comment]
as we discussed in 1-1, we use hashTransaction for getting the hash for keycard sign

**Comment by saledjenic** [comment]
> as we discussed in 1-1, we use hashTransaction for getting the hash for keycard sign

Yes, that was the point, wondering why the desktop app has never been used since it gives a lot of flexibility. Will align with the desktop app now to what/how the mobile app does and most likely we won't need this PR anymore.

**Comment by saledjenic** [comment]
@all closing this one cause it's not supported the way the mobile app does.
I've created new general endpoints in the new wallet API and once the mobile app switches to it will align the way the desktop app is dealing with these calls (basically the approach is the same just that wallet API endpoints include chain id).

For the reference, these are related PRs:
- https://github.com/status-im/status-go/pull/4393
- https://github.com/status-im/status-go/pull/4397

**Pull request closed without merging**